### PR TITLE
Small optimization to avoid unnecessary Extents merge when growing cache entry

### PR DESCRIPTION
**Pull request by alvinlin123**
**What this PR does**:

Given 2 Extents that starts at the same time, but ends at different time, it is better to put the Extent ends later (larger) at the front of the slice during sort-by-start-time time. Reason being that because the larger Extent "covers" the smaller Extents, so during the merging phase the smaller Extent simply gets dropped. 

If we put the smaller Extent before the larger Extent, then during the merge phase, the algorithm will need to figure out the overlapping point the merge the larger Extent into the smaller Extent, which is another `O(log n)` operation. 

Before:
```
Extent 1: |---|
Extent 2: |----------------|

Merging the above results in finding intersection: 

Extent 1: |---|
Extent 2: |---x-------------|

Merged:|---|-------------| (first 3 sample points are from Extent 1, rest from Extent 2)
```

After:
```
Extent 2: |----------------|
Extent 1: |---|

Merging the above is easy, simply drop Extent 1.  

Merged: |----------------| (all sample points are from Extent 1)

```

**Which issue(s) this PR fixes**:
Didn't create issue because it's an small optimization.

**Checklist**
- [ ] Tests updated
- [ ] Documentation added
- [X] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

**Review by pstibrany — APPROVED** [review]
Nice, thanks. Is it worth adding a test for this to verify intended behavior?

**Comment by alvinlin123** [comment]
> Nice, thanks. Is it worth adding a test for this to verify intended behavior?

I was pondering about the same question, but then I came to the conclusion that since this is a small optimization as long as existing tests passes then the change is functionally correct because it's not supposed to change any functional behaviour. In other words, I was thinking how I can write a test that fails before this change, but passes after change; I couldn't. If I am able to write some test that fail before this change, then something not related to this change is wrong :-) 

I also thought about refactor the `sort.Slice(...)` to a separate function, and test that function, but then this kind of test seems too testing too much into the implementation detail (testing too much of "how" rather than "what"). 

Let me know your thoughts on this as well :-)

**Comment by pstibrany** [comment]
> In other words, I was thinking how I can write a test that fails before this change, but passes after change; I couldn't.

I haven't looked into details, but I was thinking that unit test could verify that merging such extents returns original big extent, instead of two smaller ones?

**Comment by alvinlin123** [comment]
> I haven't looked into details, but I was thinking that unit test could verify that merging such extents returns original big extent, instead of two smaller ones?

Thanks for sharing your thought, I see your point. But in order to do test this we might need to modify either `Extent` or `handleHit`. Because the `mergedExtents` returned by `handleHit()` does not contain information about *how* things were merged. To the caller (the test), it's just a normal slice of `Extents`.

**Comment by pstibrany** [comment]
> To the caller (the test), it's just a normal slice of `Extents`.

Oh, I thought that the test could check if this slice is equal to input slice (= big one) or not equal (= new merged slice), but apparently slices in Go are not comparable. Perhaps it would be doable by constructing input slices with some big capacity, to be able to distinguish between them, or by checking for number of allocations (using https://golang.org/pkg/testing/#AllocsPerRun), which should be 0 if input slice was returned.

**Comment by alvinlin123** [comment]
> or by checking for number of allocations (using https://golang.org/pkg/testing/#AllocsPerRun), which should be 0 if input slice was returned unmodified.

Oh this was interesting, but I did not go with this approach because I was afraid it would make the test brittle. However, your comments gave me an idea about how to test, take a look at my new revision :-) 

I confirm that the test fails in the absence of the optimization with error that "response is nil".

**Comment by alvinlin123** [comment]
@pstibrany any more comments on this PR?

File: CHANGELOG.md [hunk 1]
```diff
@@ -25,6 +25,8 @@
 * [ENHANCEMENT] Distributor: reduce CPU and memory when an high number of errors are returned by the distributor on the write path. #3990
 * [ENHANCEMENT] Put metric before label value in the "label value too long" error message. #4018
 * [ENHANCEMENT] Allow use of `y|w|d` suffixes for duration related limits and per-tenant limits. #4044
+* [ENHANCEMENT] Query-frontend: Small optimization on top of PR #3968 to avoid unnecessary Extents merging. #4026
+
```
Comment (pracucci) [inline CHANGELOG.md:29]: Ops! This newline should be removed.

```suggestion
```

**Review by pracucci — APPROVED** [review]
Good job, LGTM. Thanks!

**Comment by pracucci** [comment]
Thanks a lot @alvinlin123 for working on this. The PR is good and I will merge it as soon as the CI pass.

For the next optimization PR (I'm sure there will be many coming ;)) I would suggest you to test the optimization with a benchmark (ideally an existing one or write a new one if none exists) in order to check its impact. 👏

**Pull request merged**